### PR TITLE
add tag tables, tag housing & public related plans

### PR DESCRIPTION
**Pull request by muamichali**
starts issue #369 

do not merge until #403 is done

### CODE REVIEW

**Review by muamichali — COMMENTED** [review]
- [x] refactor from bluebird to async await
- [x] change save function to accept multiple plan-tag thingies
- [ ] it's ok to implement in sql as long as it's readable
- [ ]  add tests for all code
- [ ] update error handling to use the project's Exception class 
- [ ] show to @TomerMe2 after implementing housing

File: server/api/controller/cron.js [hunk 1]
```diff
@@ -210,6 +208,29 @@ const sendTreeAlerts = () => {
...
+	// Before re-computing the tags of a plan, remove all previous tags for this plan.
+	
+	// TODO: Loop on the plans that need to be updated 
+	const plans = await Plan.getPlansToTag();
+	Log.info(`Processing ${plans.models.length} plans`);
+	for (const plan in plans.models) {
+		const planId = plans.models[plan].get('id');
+		await PlanTag.deletePlanTags(planId);
+		const tags = await generateTagsForPlan(planId);
+		if (tags && tags.length>0){
```
Comment (TomerMe2) [inline server/api/controller/cron.js:225]: spaces around >

File: server/api/controller/cron.js [hunk 2]
```diff
@@ -210,6 +208,29 @@ const sendTreeAlerts = () => {
...
+			await PlanTag.createPlanTags( tags);
```
Comment (TomerMe2) [inline server/api/controller/cron.js:226]: remove space before tags

File: server/api/controller/plan_tag.js [hunk 3]
```diff
@@ -0,0 +1,20 @@
+const Controller = require('../controller/controller');
+const PlanTag = require('../model/plan_tag');
+const Log = require('../lib/log');
+
+
+class PlanTagController extends Controller {
+	/**
+	 * Return person's alerts. Must be logged in.
```
Comment (TomerMe2) [inline server/api/controller/plan_tag.js:8]: I think that this comment is not related to this class

File: server/api/lib/tags/index.js [hunk 4]
```diff
@@ -0,0 +1,62 @@
+ const { tags, tagDataRules } = require('../../constants');
+ const PlanAreaChanges = require('../../../api/model/plan_area_changes');
+
+ const  generateTagsForPlan = async (planId) => {
+    let planTags = [];
```
Comment (TomerMe2) [inline server/api/lib/tags/index.js:5]: please change this let to const

File: server/api/lib/tags/index.js [hunk 5]
```diff
@@ -0,0 +1,62 @@
...
+    let dataRules = [];
```
Comment (TomerMe2) [inline server/api/lib/tags/index.js:6]: please change this let to const

File: server/api/lib/tags/index.js [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+    planTags.push( {
+        plan_id: planId,
+        tag_id: housingTag,
+        display_score: 0, /* TODO: Add the correct display score here */
+        created_by_data_rules: `[${dataRules.toString(',')}]`
+    });
+    return planTags;
+}
+
+const isTag = async (planId, tag, rule) => {
```
Comment (TomerMe2) [inline server/api/lib/tags/index.js:33]: Can you move this function into housing_tag.js os something similliar?
After discussing with the team, it's best best to keep every tag in a seperate file.

File: server/api/lib/tags/index.js [hunk 7]
```diff
@@ -0,0 +1,62 @@
+ const { tags, tagDataRules } = require('../../constants');
+ const PlanAreaChanges = require('../../../api/model/plan_area_changes');
+
+ const  generateTagsForPlan = async (planId) => {
+    let planTags = [];
+    let dataRules = [];
+    // HOUSING
+    const housingTag = tags['דיור'];
```
Comment (TomerMe2) [inline server/api/lib/tags/index.js:8]: Can you iterate on all the keys of the dictionary of tags instead of using just the housing key?
We want to make this function to work automatically on new tags that will arrive without adding code to this function. We would like this function to be 0 code maintenance for as long as we can.

I think that combining with moving the isTag function to individual files (a file per tag), you can move the function of "how to handle this dict key" to this file as well. (see the comment about moving isTag to seperate files below)

File: server/api/lib/tags/index.js [hunk 8]
```diff
@@ -0,0 +1,62 @@
...
+        tag_id: housingTag,
+        display_score: 0, /* TODO: Add the correct display score here */
+        created_by_data_rules: `[${dataRules.toString(',')}]`
+    });
+    return planTags;
+}
+
+const isTag = async (planId, tag, rule) => {
+    try {
+        const result = await PlanAreaChanges.byPlanAndUsage(planId, tagDataRules[rule].usage);
```
Comment (TomerMe2) [inline server/api/lib/tags/index.js:35]: Please add spaces around commans or operators in this function

File: server/api/model/plan_area_changes.js [hunk 9]
```diff
@@ -0,0 +1,55 @@
+const Model = require('./base_model');
+const Log = require('../lib/log');
+const Exception = require('./exception');
+const { Bookshelf, Knex } = require('../service/database');
+const {	tags, tagDataRules } = require('../constants');
+
+class PlanAreaChanges extends Model {
```
Comment (TomerMe2) [inline server/api/model/plan_area_changes.js:7]: After the merge of the other PR, you sohuld merge it to this branch. I put some comments about this file on the other PR.

**Review by TomerMe2 — CHANGES_REQUESTED** [review]
Great work! Thank you!
We should make the process of creating the tags more generic in my opinion. It will be better because we will have to write less code when introducing new tags and it will be easier to implement new tags.

File: server/api/lib/tags/index.js (hunk 4 above)
Comment (muamichali) [reply server/api/lib/tags/index.js:5]: but I add things to it, so it's not a const...

File: server/api/lib/tags/index.js (hunk 5 above)
Comment (muamichali) [reply server/api/lib/tags/index.js:6]: but I add things to it, so it's not a const...

File: server/api/lib/tags/index.js (hunk 4 above)
Comment (TomerMe2) [reply server/api/lib/tags/index.js:5]: Cost and let are referring to the reference. The reference is set once at your code, on deceleration time.
After that you are mutating the object, but don't change the reference. In that scenario, const is preferable. 

Legit code using const:
```
const a = [];
a.push(2);  //mutating the object
```
Non legit code that will fail:
```
const a = [];
a = [2];   // mutate the reference
```

File: server/api/model/plan_area_changes.js (hunk 9 above)
Comment (muamichali) [reply server/api/model/plan_area_changes.js]: Done.

File: server/api/lib/tags/housing.js [hunk 10]
```diff
@@ -0,0 +1,36 @@
...
+const { isTagByUsageAddition} = require('../tags/utils');
+
+const isHousing = async (planId, tag) => {  
+	const HOUSING_BY_AREA = 'housingByArea';
+	const HOUSING_BY_UNITS = 'housingByUnits';
+	const planTags = [];
+	const dataRules = [];
+	
+	// check housing by area
+	const housingByArea = await isTagByUsageAddition(planId,tag,HOUSING_BY_AREA);
```
Comment (TomerMe2) [inline server/api/lib/tags/housing.js:11]: Please add spaces between operands

File: server/api/lib/tags/housing.js [hunk 11]
```diff
@@ -0,0 +1,36 @@
...
+	
+	// check housing by area
+	const housingByArea = await isTagByUsageAddition(planId,tag,HOUSING_BY_AREA); 
+	if (housingByArea) {
+		dataRules.push(
+			housingByArea.created_by_data_rules
+		);
+	}
+	// check housing by units
+	const housingByUnits = await isTagByUsageAddition(planId,tag,HOUSING_BY_UNITS);
```
Comment (TomerMe2) [inline server/api/lib/tags/housing.js:18]: Please add spaces between operands

File: server/api/lib/tags/utils.js [hunk 12]
```diff
@@ -0,0 +1,33 @@
+const { tagDataRules } = require('../../constants');
+const PlanAreaChanges = require('../../../api/model/plan_area_changes');
+
+const isTagByUsageAddition = async (planId, tag, rule) => {
+	try {
+		const result = await PlanAreaChanges.byPlanAndUsage(planId, tagDataRules[rule].usage);
+		if (result && result.models && result.models[0]) {
+			const changeToApprovedState = result.models[0].attributes.change_to_approved_state;
+			if ( (changeToApprovedState.length>1) &&
```
Comment (TomerMe2) [inline server/api/lib/tags/utils.js:9]: please add spaces between operators and operands

**Review by TomerMe2 — CHANGES_REQUESTED** [review]
Great work! Thanks!

File: server/api/constants.js [hunk 13]
```diff
@@ -13,3 +13,21 @@ exports.personTypes = {
 	'5': 'יזם',
 	'6': 'אחר'
 };
+
+exports.tags = [
+	{'tagName':'דיור' , 'tagId':'1', 'functionName': 'isHousing'}
```
Comment (GalGend) [inline server/api/constants.js:18]: When using objects, no need to add `'` for keys, no need to change but for next tine this can also work:
```
{tagName:'housing'....}
```

File: server/api/lib/tags/housing.js [hunk 14]
```diff
@@ -0,0 +1,36 @@
...
+		planTags.push( {
+			plan_id: planId,
+			tag_id: tag,
+			display_score: 0, /* TODO: Add the correct display score here */
+			created_by_data_rules: `[${dataRules.toString(',')}]`
+		});
+	return planTags;
+}
+
+module.exports = {
```
Comment (GalGend) [inline server/api/lib/tags/housing.js:34]: General structure for tags:

We have a folder of tags= housing, employment, commerce, etc. 
Each file (housing.js) should export one function- `doesApply` or `doesTagApply` or any name you find suitable. Then, we have an index.js file with all tags, their label (as done in constants.js, but in our tag folder) and the isApply function.

Comment (GalGend) [reply server/api/lib/tags/housing.js:34]: Then, the code that will run the tag-extraction can access all rules using the index.js file, and shouldn't know it explicitly

File: server/api/controller/cron.js [hunk 15]
```diff
@@ -210,6 +208,29 @@ const sendTreeAlerts = () => {
...
 
+const updatePlanTags = async () => {
+	let start = Number(Date.now());
+	Log.info('Re-creating plan tags');
+	let tagCounter = 0;
+	// Re-compute the tags of a plan if the last update time of the plan is after the last update time of the tags of this plan.
+	// Before re-computing the tags of a plan, remove all previous tags for this plan.
+	
+	// TODO: Loop on the plans that need to be updated 
+	const plans = await Plan.getPlansToTag();
```
Comment (GalGend) [inline server/api/controller/cron.js:219]: Do we set limits on how many plans we fetch to extract tags from?

**Review by GalGend — COMMENTED** [review]
Hi, looks good :) Left a general comment on how we should structure our tags and rules files.

File: server/api/lib/tags/housing.js (hunk 14 above)
Comment (muamichali) [reply server/api/lib/tags/housing.js:36]: Sounds good. I refactored it based on this feedback in https://github.com/meirim-org/meirim/pull/389/commits/7583ef1a2cfe5292725020c9e3e41870d8323ba8 PLMK if it looks good!

File: server/api/controller/cron.js (hunk 15 above)
Comment (muamichali) [reply server/api/controller/cron.js:219]: After discussing with @TomerMe2 I am going to implement the following logic in getPlansToTag: 
1. add an indicator of when we last updated the tags to the plan table
2. update tags last updated indicator when we update tags
3. compare it to when the plan was last updated in this function. 

Do you want to also set a max number of records?

Comment (GalGend) [reply server/api/controller/cron.js:219]: Hi Michal
As for now, Tomer explained to me that the tag generation is an easy task, and doesn't require loads of processing. Anyways, I think it would be better if we do manage the number of plans we generate tags for, as our server has no resources

File: server/api/controller/plan_tag.js [hunk 16]
```diff
@@ -0,0 +1,20 @@
...
+
+
+class PlanTagController extends Controller {
+	/**
+	 * Return tags for this plan.
+	 * @param {IncomingRequest} req
+	 */
+	byPlan (req) {
+		return this.model.byPlan(req.params.plan_id).then(collection => {
+			Log.debug(this.tableName, 'Get tag list', req.params.plan_id);
```
Comment (GalGend) [inline server/api/controller/plan_tag.js:13]: What fields of the tag it returns? Its ID or populate it with the tag table to get its labels?

File: server/api/lib/tags/utils.js [hunk 17]
```diff
@@ -0,0 +1,33 @@
+const { tagDataRules } = require('../../constants');
+const PlanAreaChanges = require('../../../api/model/plan_area_changes');
+
+const isTagByUsageAddition = async (planId, tag, rule) => {
+	try {
+		const result = await PlanAreaChanges.byPlanAndUsage(planId, tagDataRules[rule].usage);
+		if (result && result.models && result.models[0]) {
```
Comment (GalGend) [inline server/api/lib/tags/utils.js:7]: Also, I think it'd be better if the rule description should be done in the housing file, and the isTagByUsageAddition function should also return true/ false.

For example, send to the function `isTagByUsageAddition` what it needs to check, and it'll return true/ false. Then the housing will add it .

In this case, the `isTagByUsageAddition` function does not know the logic of the rules, and only performs what it's meant to do - check a specific land use change. Also, it will not require sending it the tag info

```
if (housingByUnits) {
		dataRules.push(
			created_by_data_rules: {   
						tag_id: tag, 
						created_by_data_rules: `{rule:'${tagDataRules[rule].description}',detail:'adds ${changeToApprovedState} ${tagDataRules[rule].usage}'}`
					}
		);
	}

```

File: server/api/controller/plan_tag.js (hunk 16 above)
Comment (muamichali) [reply server/api/controller/plan_tag.js:13]: It returns the tag name and tag display name from the Tag table. See https://github.com/muamichali/meirim/blob/7583ef1a2cfe5292725020c9e3e41870d8323ba8/server/api/model/plan_tag.js#L37

File: server/api/controller/cron.js (hunk 15 above)
Comment (muamichali) [reply server/api/controller/cron.js:219]: CC @TomerMe2 when you come back. I am talking to @GalGend and she says that we don't handle plan updates yet, so maybe this is not necessary yet?

**Review by TomerMe2 — APPROVED** [review]
